R/textmodel_nb.R

```diff
@@ -331,3 +331,80 @@ colNorm <- function(x) {
 print.predict.textmodel_nb <- function(x, ...) {
     print(unclass(x))
 }
+
+
+#' Compute the information gain of features from a fitted \link{textmodel_nb}
+#' object
+#' 
+#' Computes the information gain of features from a fitted \link{textmodel_nb}
+#' object.  See  
+#' @param model a \link{textmodel_nb} object
+#' @param class_specific  whether to return class specific information gain
+#'   scores or not. Defaults to \code{TRUE}.
+#' @param base the base of the logarithmic function to be used. Defaults to 2.
+#' @references 
+#' Jurafsky, Daniel and James H. Martin.  (2018).
+#' \href{https://web.stanford.edu/~jurafsky/slp3/4.pdf}{\emph{Speech and
+#' Language Processing}}.   Stanford University typescript.  Draft of September
+#' 23.
+#' @keywords textmodel textmodel_nb
+#' @examples 
+#' txt <- c(d1 = "Chinese Beijing Chinese",
+#'          d2 = "Chinese Chinese Shanghai",
+#'          d3 = "Chinese Macao",
+#'          d4 = "Tokyo Japan Chinese",
+#'          d5 = "Chinese Chinese Chinese Tokyo Japan")
+#' trainingset <- dfm(txt, tolower = FALSE)
+#' trainingclass <- factor(c("Y", "Y", "Y", "N", NA), ordered = TRUE)
+#' tmod <- textmodel_nb(trainingset, y = trainingclass, prior = "docfreq")
+#' gain.textmodel_nb(tmod)
+#' gain.textmodel_nb(tmod, class_specific = FALSE)
+#' @export
+gain.textmodel_nb <- function(model, base = 2, class_specific = TRUE) {
+    if (!("textmodel_nb" %in% class(model))) stop("model must be a textmodel_nb object")
+    if (class(class_specific) != "logical") stop("class_specific must be a logical (TRUE or FALSE)")
+
+    # The formula for entropy of word w is
+    # - summation over all classes ( P(class) * log P(class))  (Term 1)
+    # + P(w) * summation over all classes (P(class |w ) * log(P(class|w)) ) (Term 2)
+    # + P(w') * summation over all classes (P(class | w') * log(P(class|w')) ) (Term 3)
+
+    # Term 1: Initial Entropy
+    # Get summation over all classes of  P(class) * log P(class)
+    initial_entropy <- apply(data.frame(model$Pc), MARGIN = 2, function(x) -1 * x * log(x, base = base))
```

Comment:
Not sure why you are coercing these to data.frames when keeping them as matrix quantities will work just fine.

Reply:
i initially coerced to data.frame because I ran into the following error:
```r
 Error in apply(conditional_entropy, 1, `*`, object$Pw) : 
  dim(X) must have a positive length 
```
but I have since fixed this by using `sapply`